rust/worker/src/index/metadata/types.rs

```diff
@@ -49,115 +51,166 @@ pub(crate) enum MetadataIndexWriter {
     // and the expected case is much less than that.
     F32MetadataIndexWriter(
         BlockfileWriter,
+        Option<MetadataIndexReader<'me>>,
         Arc<Mutex<HashMap<String, Vec<(f32, RoaringBitmap)>>>>,
     ),
     BoolMetadataIndexWriter(
         BlockfileWriter,
+        Option<MetadataIndexReader<'me>>,
         Arc<Mutex<HashMap<String, HashMap<bool, RoaringBitmap>>>>,
     ),
 }
 
-impl MetadataIndexWriter {
-    pub fn new_string(init_blockfile_writer: BlockfileWriter) -> Self {
+impl<'me> MetadataIndexWriter<'me> {
+    pub fn new_string(
+        init_blockfile_writer: BlockfileWriter,
+        string_metadata_index_reader: Option<MetadataIndexReader<'me>>,
+    ) -> Self {
         MetadataIndexWriter::StringMetadataIndexWriter(
             init_blockfile_writer,
+            string_metadata_index_reader,
             Arc::new(Mutex::new(HashMap::new())),
         )
     }
 
-    pub fn new_u32(init_blockfile_writer: BlockfileWriter) -> Self {
+    pub fn new_u32(
+        init_blockfile_writer: BlockfileWriter,
+        int_metadata_index_reader: Option<MetadataIndexReader<'me>>,
+    ) -> Self {
         MetadataIndexWriter::U32MetadataIndexWriter(
             init_blockfile_writer,
+            int_metadata_index_reader,
             Arc::new(Mutex::new(HashMap::new())),
         )
     }
 
-    pub fn new_f32(init_blockfile_writer: BlockfileWriter) -> Self {
+    pub fn new_f32(
+        init_blockfile_writer: BlockfileWriter,
+        f32_metadata_index_reader: Option<MetadataIndexReader<'me>>,
+    ) -> Self {
         MetadataIndexWriter::F32MetadataIndexWriter(
             init_blockfile_writer,
+            f32_metadata_index_reader,
             Arc::new(Mutex::new(HashMap::new())),
         )
     }
 
-    pub fn new_bool(init_blockfile_writer: BlockfileWriter) -> Self {
+    pub fn new_bool(
+        init_blockfile_writer: BlockfileWriter,
+        bool_metadata_index_reader: Option<MetadataIndexReader<'me>>,
+    ) -> Self {
         MetadataIndexWriter::BoolMetadataIndexWriter(
             init_blockfile_writer,
+            bool_metadata_index_reader,
             Arc::new(Mutex::new(HashMap::new())),
         )
     }
 
-    fn look_up_key_and_populate_uncommitted_rbms(
+    async fn look_up_key_and_populate_uncommitted_rbms(
```

Comment:
nit: can you add a small description of this function so that others working on it in future get context immediately